### ISSUE TITLE
CL v17.2.0 might have issues on payee notifications, Duplicate transfer ID check as well

### ISSUE THREAD

**Issue by sri-miriyala** [mojaloop/project#3527]
**Summary**:
When running perf tests in the EC2 virtual machine, I am seeing the below error when running a transfer. And the DB records of a particular transfer ID is showing 6 records instead of 4. 
Pls make sure this value is set 'false' : SEND_TRANSFER_CONFIRMATION_TO_PAYEE=false
mysql> select * from transferStateChange where transferId="a9721abd-7495-437a-a974-d3165ec75728"
    -> ;
+-----------------------+--------------------------------------+------------------+-----------------------------+---------------------+
| transferStateChangeId | transferId                           | transferStateId  | reason                      | createdDate         |
+-----------------------+--------------------------------------+------------------+-----------------------------+---------------------+
|                494604 | a9721abd-7495-437a-a974-d3165ec75728 | RECEIVED_PREPARE | NULL                        | 2023-09-15 14:02:30 |
|                494605 | a9721abd-7495-437a-a974-d3165ec75728 | RESERVED         | NULL                        | 2023-09-15 14:02:30 |
|                494606 | a9721abd-7495-437a-a974-d3165ec75728 | RECEIVED_FULFIL  | NULL                        | 2023-09-15 14:02:32 |
|                494607 | a9721abd-7495-437a-a974-d3165ec75728 | COMMITTED        | NULL                        | 2023-09-15 14:02:33 |
|                494608 | a9721abd-7495-437a-a974-d3165ec75728 | ABORTED_REJECTED | Transfer in incorrect state | 2023-09-15 14:02:33 |
|                494609 | a9721abd-7495-437a-a974-d3165ec75728 | ABORTED_REJECTED | Transfer in incorrect state | 2023-09-15 14:02:33 |
+-----------------------+--------------------------------------+------------------+-----------------------------+---------------------+

Possible issues:
1. Duplicate check logic on Prepare-Handler doesn’t seem to be catching the duplicate transfers, i.e. we shouldn’t be getting a position change event being processed by the Position Handler
2. The ml-api-adapter notification logic for the SEND_TRANSFER_CONFIRMATION_TO_PAYEE=false doesn't seem to be working anymore.
3. 
**Severity**:
(Low, Medium, High)

**Priority**:
(Critical, Medium, Low)

**Expected Behavior**
1. The transfer State Change table should be showing only 4 records per transfer ID with "COMMITTED" as the last state on a successful transfer
2. If the same transfer ID is triggerred again, it should end up with a duplicate transfer ID error. and it should not be showing a different status in the DB.

**Steps to Reproduce**
  1. Connect to EC2 machine
  2. make sure all services are up and running like ml-core, monitoring, k6
  3. Run single transfer using this command : 
  env K6_SCRIPT_CONFIG_FILE_NAME=fspiopSingleTransfer.json docker compose --project-name load -f docker-compose-load.yml up
  5. check the DB transferStatusChange table, for that transfer ID.
  6. to check logs use below command:
docker compose --project-name ml-core -f docker-compose-perf.yml --profile transfers-test --profile 8dfsp --profile ttk-provisioning-transfers logs -f --since 10s

**Specifications**
  - Component (if known):
  - Version: CL v.17.2.0, ML_API version: 14.0.4
  - Platform:
  - Subsystem:
  - Type of testing:
  - Bug found/raised by: @sri-miriyala 

**Notes**:

- Severity when opened:
- Priority when opened:

**PRs**
- [ ] https://github.com/mojaloop/object-store-lib/pull/47
- [ ] https://github.com/mojaloop/central-ledger/pull/980

**Comment by PaulGregoryBaker** [comment]
CL v17.0.4 does not have this error.

**Comment by PaulGregoryBaker** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/mojaloop-project-59edee71d1407922110cf083/issues/mojaloop/project/3527?planning-poker) with Zenhub @aaronreynoza @kleyow @oderayi @sri-miriyala @vijayg10

**Comment by oderayi** [comment]
Issues observed:
1. The latest kafka library fails to auto create topics, therefore kafka provisioning needs to be completed successfully before attempts are made to start ML-API-Adapter and Central Ledger services or before any attempt is made to consume messages. 
2. The kafka provisioning script in CL had a bug. One of the topics (topic-bulk-get) had invalid character
3. The object-store-lib is broken, the `mongoose.connect` method signature has changed in the latest version of mongoose.  This has been fixed and released in version `12.0.2` of object-store-lib. CL should be updated to this version.

**Comment by oderayi** [comment]
With `central-services-stream:v11.0.0` (downgrade):
```
mysql> select * from transferStateChange;
+-----------------------+--------------------------------------+------------------+--------+---------------------+
| transferStateChangeId | transferId                           | transferStateId  | reason | createdDate         |
+-----------------------+--------------------------------------+------------------+--------+---------------------+
|                     1 | 87c0f49d-2f6b-4344-a532-19a6eb4669f5 | RECEIVED_PREPARE | string | 2023-09-28 11:18:43 |
|                     2 | 87c0f49d-2f6b-4344-a532-19a6eb4669f5 | RESERVED         | string | 2023-09-28 11:18:43 |
|                     3 | 87c0f49d-2f6b-4344-a532-19a6eb4669f5 | RECEIVED_FULFIL  | string | 2023-09-28 11:18:43 |
|                     4 | 87c0f49d-2f6b-4344-a532-19a6eb4669f5 | COMMITTED        | string | 2023-09-28 11:18:43 |
|                     5 | 3adab467-6b70-4a99-9655-1d765f0ad872 | RECEIVED_PREPARE | string | 2023-09-28 11:18:44 |
|                     6 | 3adab467-6b70-4a99-9655-1d765f0ad872 | RESERVED         | string | 2023-09-28 11:18:44 |
|                     7 | 3adab467-6b70-4a99-9655-1d765f0ad872 | RECEIVED_FULFIL  | string | 2023-09-28 11:18:44 |
|                     8 | 3adab467-6b70-4a99-9655-1d765f0ad872 | COMMITTED        | string | 2023-09-28 11:18:44 |
|                     9 | dd8e5db4-577c-45f1-8b0f-89553aaf876c | RECEIVED_PREPARE | string | 2023-09-28 11:19:00 |
|                    10 | dd8e5db4-577c-45f1-8b0f-89553aaf876c | RESERVED         | string | 2023-09-28 11:19:00 |
|                    11 | dd8e5db4-577c-45f1-8b0f-89553aaf876c | RECEIVED_FULFIL  | string | 2023-09-28 11:19:00 |
|                    12 | dd8e5db4-577c-45f1-8b0f-89553aaf876c | COMMITTED        | string | 2023-09-28 11:19:00 |
|                    13 | e9dbaec3-bc0f-4596-8a17-2796ebc9507a | RECEIVED_PREPARE | string | 2023-09-28 11:19:02 |
|                    14 | e9dbaec3-bc0f-4596-8a17-2796ebc9507a | RESERVED         | string | 2023-09-28 11:19:02 |
|                    15 | e9dbaec3-bc0f-4596-8a17-2796ebc9507a | RECEIVED_FULFIL  | string | 2023-09-28 11:19:02 |
|                    16 | e9dbaec3-bc0f-4596-8a17-2796ebc9507a | COMMITTED        | string | 2023-09-28 11:19:02 |
|                    17 | 9aa5e84a-78c4-462c-9f0f-c230854b59c5 | RECEIVED_PREPARE | string | 2023-09-28 11:19:17 |
|                    18 | 9aa5e84a-78c4-462c-9f0f-c230854b59c5 | RESERVED         | string | 2023-09-28 11:19:17 |
|                    19 | 9aa5e84a-78c4-462c-9f0f-c230854b59c5 | RECEIVED_FULFIL  | string | 2023-09-28 11:19:17 |
|                    20 | 9aa5e84a-78c4-462c-9f0f-c230854b59c5 | COMMITTED        | string | 2023-09-28 11:19:17 |
|                    21 | c8c16725-061e-4255-8c0d-5bbca189ea68 | RECEIVED_PREPARE | string | 2023-09-28 11:19:19 |
|                    22 | c8c16725-061e-4255-8c0d-5bbca189ea68 | RESERVED         | string | 2023-09-28 11:19:19 |
|                    23 | c8c16725-061e-4255-8c0d-5bbca189ea68 | RECEIVED_FULFIL  | string | 2023-09-28 11:19:19 |
|                    24 | c8c16725-061e-4255-8c0d-5bbca189ea68 | COMMITTED        | string | 2023-09-28 11:19:19 |
|                    25 | b2eb9224-698a-49a5-b043-0e66e4839a38 | RECEIVED_PREPARE | string | 2023-09-28 11:19:33 |
|                    26 | b2eb9224-698a-49a5-b043-0e66e4839a38 | RESERVED         | string | 2023-09-28 11:19:33 |
|                    27 | b2eb9224-698a-49a5-b043-0e66e4839a38 | RECEIVED_FULFIL  | string | 2023-09-28 11:19:33 |
|                    28 | b2eb9224-698a-49a5-b043-0e66e4839a38 | COMMITTED        | string | 2023-09-28 11:19:33 |
|                    29 | 584de4bd-55be-42aa-b299-196e875fdc47 | RECEIVED_PREPARE | string | 2023-09-28 11:19:35 |
|                    30 | 584de4bd-55be-42aa-b299-196e875fdc47 | RESERVED         | string | 2023-09-28 11:19:35 |
|                    31 | 584de4bd-55be-42aa-b299-196e875fdc47 | RECEIVED_FULFIL  | string | 2023-09-28 11:19:35 |
|                    32 | 584de4bd-55be-42aa-b299-196e875fdc47 | COMMITTED        | string | 2023-09-28 11:19:35 |
|                    33 | 60455f5d-48e7-499b-80fe-08c1d767e716 | RECEIVED_PREPARE | string | 2023-09-28 11:19:50 |
|                    34 | 60455f5d-48e7-499b-80fe-08c1d767e716 | RESERVED         | string | 2023-09-28 11:19:50 |
|                    35 | 60455f5d-48e7-499b-80fe-08c1d767e716 | RECEIVED_FULFIL  | string | 2023-09-28 11:19:50 |
|                    36 | 60455f5d-48e7-499b-80fe-08c1d767e716 | COMMITTED        | string | 2023-09-28 11:19:50 |
|                    37 | 1de8a3ee-9082-4908-99be-b554c0753c0f | RECEIVED_PREPARE | string | 2023-09-28 11:19:52 |
|                    38 | 1de8a3ee-9082-4908-99be-b554c0753c0f | RESERVED         | string | 2023-09-28 11:19:52 |
|                    39 | 1de8a3ee-9082-4908-99be-b554c0753c0f | RECEIVED_FULFIL  | string | 2023-09-28 11:19:52 |
|                    40 | 1de8a3ee-9082-4908-99be-b554c0753c0f | COMMITTED        | string | 2023-09-28 11:19:52 |
|                    41 | fb026577-1933-403d-8c65-7a87b9dbc9be | RECEIVED_PREPARE | string | 2023-09-28 11:20:08 |
|                    42 | fb026577-1933-403d-8c65-7a87b9dbc9be | RESERVED         | string | 2023-09-28 11:20:08 |
|                    43 | fb026577-1933-403d-8c65-7a87b9dbc9be | RECEIVED_FULFIL  | string | 2023-09-28 11:20:08 |
|                    44 | fb026577-1933-403d-8c65-7a87b9dbc9be | COMMITTED        | string | 2023-09-28 11:20:08 |
|                    45 | 617cc069-4c07-4735-8f5e-fd0dcacfd9c9 | RECEIVED_PREPARE | string | 2023-09-28 11:20:09 |
|                    46 | 617cc069-4c07-4735-8f5e-fd0dcacfd9c9 | RESERVED         | string | 2023-09-28 11:20:09 |
|                    47 | 617cc069-4c07-4735-8f5e-fd0dcacfd9c9 | RECEIVED_FULFIL  | string | 2023-09-28 11:20:09 |
|                    48 | 617cc069-4c07-4735-8f5e-fd0dcacfd9c9 | COMMITTED        | string | 2023-09-28 11:20:09 |
|                    49 | 46a0b665-e435-4dad-89bb-0291b394dd0f | RECEIVED_PREPARE | string | 2023-09-28 11:20:24 |
|                    50 | 46a0b665-e435-4dad-89bb-0291b394dd0f | RESERVED         | string | 2023-09-28 11:20:24 |
|                    51 | 46a0b665-e435-4dad-89bb-0291b394dd0f | RECEIVED_FULFIL  | string | 2023-09-28 11:20:24 |
|                    52 | 46a0b665-e435-4dad-89bb-0291b394dd0f | COMMITTED        | string | 2023-09-28 11:20:24 |
|                    53 | b9d63011-0a39-403e-9c06-57e1f5eaac51 | RECEIVED_PREPARE | string | 2023-09-28 11:20:26 |
|                    54 | b9d63011-0a39-403e-9c06-57e1f5eaac51 | RESERVED         | string | 2023-09-28 11:20:26 |
|                    55 | b9d63011-0a39-403e-9c06-57e1f5eaac51 | RECEIVED_FULFIL  | string | 2023-09-28 11:20:26 |
|                    56 | b9d63011-0a39-403e-9c06-57e1f5eaac51 | COMMITTED        | string | 2023-09-28 11:20:26 |
|                    57 | 30fe3ff4-6831-4867-9dbc-b53c6fe5e2b5 | RECEIVED_PREPARE | string | 2023-09-28 11:20:45 |
|                    58 | 30fe3ff4-6831-4867-9dbc-b53c6fe5e2b5 | RESERVED         | string | 2023-09-28 11:20:45 |
|                    59 | 30fe3ff4-6831-4867-9dbc-b53c6fe5e2b5 | RECEIVED_FULFIL  | string | 2023-09-28 11:20:45 |
|                    60 | 30fe3ff4-6831-4867-9dbc-b53c6fe5e2b5 | COMMITTED        | string | 2023-09-28 11:20:45 |
|                    61 | 190a03e6-3478-4b10-9d2d-3b67a7af68b5 | RECEIVED_PREPARE | string | 2023-09-28 11:20:47 |
|                    62 | 190a03e6-3478-4b10-9d2d-3b67a7af68b5 | RESERVED         | string | 2023-09-28 11:20:47 |
|                    63 | 190a03e6-3478-4b10-9d2d-3b67a7af68b5 | RECEIVED_FULFIL  | string | 2023-09-28 11:20:47 |
|                    64 | 190a03e6-3478-4b10-9d2d-3b67a7af68b5 | COMMITTED        | string | 2023-09-28 11:20:47 |
|                    65 | 5cc17e91-8696-4b08-8181-9ae353be3f86 | RECEIVED_PREPARE | string | 2023-09-28 11:21:21 |
|                    66 | 5cc17e91-8696-4b08-8181-9ae353be3f86 | RESERVED         | string | 2023-09-28 11:21:21 |
|                    67 | 5cc17e91-8696-4b08-8181-9ae353be3f86 | RECEIVED_FULFIL  | string | 2023-09-28 11:21:21 |
|                    68 | 5cc17e91-8696-4b08-8181-9ae353be3f86 | COMMITTED        | string | 2023-09-28 11:21:21 |
|                    69 | 23783bc0-49a8-48e9-8143-a3c300e8eba0 | RECEIVED_PREPARE | string | 2023-09-28 11:21:23 |
|                    70 | 23783bc0-49a8-48e9-8143-a3c300e8eba0 | RESERVED         | string | 2023-09-28 11:21:23 |
|                    71 | 23783bc0-49a8-48e9-8143-a3c300e8eba0 | RECEIVED_FULFIL  | string | 2023-09-28 11:21:23 |
|                    72 | 23783bc0-49a8-48e9-8143-a3c300e8eba0 | COMMITTED        | string | 2023-09-28 11:21:23 |
+-----------------------+--------------------------------------+------------------+--------+---------------------+

```

**Comment by oderayi** [comment]
Without downgrading:
These errors are observed in `central-ledger` terminal:

```
2023-09-28T11:46:45.827Z - error: "Consumer::onError()[topics='topic-transfer-position'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.859Z - error: "Consumer::onError()[topics='topic-transfer-prepare'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.859Z - error: "Consumer::onError()[topics='topic-transfer-fulfil'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.870Z - error: "Consumer::onError()[topics='topic-transfer-get'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.899Z - error: "Consumer::onError()[topics='topic-bulk-fulfil'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.903Z - error: "Consumer::onError()[topics='topic-bulk-get'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.912Z - error: "Consumer::onError()[topics='topic-bulk-prepare'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.918Z - error: "Consumer::onError()[topics='topic-bulk-processing'] - Error: Broker: Unknown topic or partition\n    at Function.createLibrdkafkaError [as create] (/opt/app/node_modules/node-rdkafka/lib/error.js:456:10)\n    at /opt/app/node_modules/node-rdkafka/lib/kafka-consumer.js:468:29)"
2023-09-28T11:46:45.928Z - error: "Consumer::consume() - onRecursive - error Error: Broker: Unknown topic or partition"

```

**Comment by oderayi** [comment]
Bug could not be reproduced except for performance related bugs which have been fixed in #3529 
@sri-miriyala to re-run tests and confirm before closing this story.

**Comment by sri-miriyala** [comment]
With the latest versions of the CL, 17.1.2 (which has the fix for transfer lag issue #3529)
we are not seeing this issue. Below screenshot shows the transfers records from trasnferStateChange table from DB where there is no state of "ABORTED" and no reason "Transfer in incorrect state". We ran multiple times with 10 min each iteration.
This shows this issue also fixed.

<img width="812" alt="Screenshot 2023-10-04 at 12 02 37 PM" src="https://github.com/mojaloop/project/assets/38143240/2260dbb0-b112-45c1-9084-86414257cd15">
<img width="517" alt="Screenshot 2023-10-04 at 12 00 42 PM" src="https://github.com/mojaloop/project/assets/38143240/77667da4-3ba0-466d-9b41-33abdb25b2bf">


Hence the issue is closed.
